eosio.unregd/eosio.unregd.cpp

```diff
@@ -21,6 +21,103 @@ void unregd::add(const ethereum_address& ethereum_address, const asset& balance)
   });
 }
 
+/**
+ * Register an EOS account using the stored information (address/balance) verifying an ETH signature
+ */
+void unregd::regaccount(const bytes& signature, const string& account, const eosio::public_key& eos_pubkey) {
+
+  eosio_assert(signature.size() == 66, "Invalid signature");
+  eosio_assert(account.size() == 12, "Invalid account length");
+
+  // Verify that the destination account name is valid
+  for(const auto& c : account) {
+    if(!((c >= 'a' && c <= 'z') || (c >= '1' && c <= '5')))
+      eosio_assert(false, "Invalid account name");
+  }
+
+  auto naccount = string_to_name(account.c_str());
+
+  // Verify that the account does not exists
+  eosio_assert(!is_account(naccount), "Account already exists");
+
+  // Calculate message hash based on current TX block num/prefix
+  char* message = (char*)malloc(64);
+  sprintf(message, "%u,%u", tapos_block_num(), tapos_block_prefix());
```

Comment:
As @wanderingbort mentioned, it is critical for the digest to commit also to the public key and ideally the account as well. Otherwise an attacker could potentially steal the funds of a victim who was in the process of registering using `unregd:regaccount`.

Also, have you considered using a struct to store `tapos_block_num()`, `tapos_block_prefix()`, `naccount`, and `eos_pubkey`, then serializing that struct into a buffer which is then sha256 hashed to give you the `digest`? That way you can remove this dependency on `sprintf` which would reduce contract size.